File: src/util/ResultCE.fs
module ResultCE

type ResultBuilder() =
    member _.Return(x) = Ok x
    member _.ReturnFrom(m) = m
    member _.Bind(m, f) =
        match m with
        | Ok x -> f x
        | Error e -> Error e
    member _.Zero() = Ok ()
    member _.Combine(m1, m2) =
        match m1 with
        | Ok () -> m2 ()
        | Error e -> Error e
    member _.Delay(f) = f
    member _.Run(f) = f ()

let result = ResultBuilder()

let resultTraverse f items =
    List.foldBack (fun item acc ->
        match acc with
        | Error e -> Error e
        | Ok xs ->
            match f item with
            | Ok x -> Ok (x :: xs)
            | Error e -> Error e) items (Ok [])

let resultFold f state items =
    List.fold (fun acc item ->
        match acc with
        | Error e -> Error e
        | Ok s -> f s item) (Ok state) items

File: src/util/ResultCE.rs
//! A direct Rust analogue of the F# result computation-expression builder.

#[derive(Clone, Copy, Debug, Default)]
pub struct ResultBuilder;

impl ResultBuilder {
    pub fn return_<T, E>(&self, value: T) -> Result<T, E> { Ok(value) }
    pub fn return_from<T, E>(&self, result: Result<T, E>) -> Result<T, E> { result }
    pub fn bind<T, U, E>(&self, result: Result<T, E>, f: impl FnOnce(T) -> Result<U, E>) -> Result<U, E> {
        result.and_then(f)
    }
    pub fn zero<E>(&self) -> Result<(), E> { Ok(()) }
    pub fn combine<T, E>(&self, first: Result<(), E>, second: impl FnOnce() -> Result<T, E>) -> Result<T, E> {
        first.and_then(|()| second())
    }
    pub fn delay<T, E, F: FnOnce() -> Result<T, E>>(&self, f: F) -> F { f }
    pub fn run<T, E>(&self, f: impl FnOnce() -> Result<T, E>) -> Result<T, E> { f() }
}

pub const RESULT: ResultBuilder = ResultBuilder;
